code/datums/elements/pathfinder.dm

```diff
@@ -37,33 +37,37 @@ stutter_step: a prob() chance to go left or right of the mob's direction towards
 	stutter_step_prob[target] = stutter_step
 
 /datum/element/pathfinder/process()
-	for(var/mob in distances_to_maintain)
-		var/mob/mob_to_process = mob
-		if(!mob_to_process.canmove || mob_to_process.stat == DEAD)
+	for(var/mob/mob_to_process AS in distances_to_maintain)
+		if(!mob_to_process.canmove)
 			continue
 
 		//Okay it can actually physically move, but has it moved too recently?
-		if(world.time <= mob_to_process.last_move_time + mob_to_process.cached_multiplicative_slowdown || mob_to_process.do_actions)
+		if(world.time <= mob_to_process.next_move || mob_to_process.do_actions)
 			continue
-
+		var/step_dir
 		if(get_dist(mob_to_process, atoms_to_walk_to[mob_to_process]) == distances_to_maintain[mob_to_process])
 			SEND_SIGNAL(mob_to_process, COMSIG_STATE_MAINTAINED_DISTANCE)
-			if(!(get_dir(mob_to_process, atoms_to_walk_to[mob_to_process]))) //We're right on top, move out of it
-				if(!step(mob_to_process, pick(CARDINAL_ALL_DIRS)))
-					SEND_SIGNAL(mob_to_process, COMSIG_OBSTRUCTED_MOVE)
+			if(!get_dir(mob_to_process, atoms_to_walk_to[mob_to_process])) //We're right on top, move out of it
+				step_dir = pick(CARDINAL_ALL_DIRS)
+				if(!mob_to_process.Move(get_step(mob_to_process, step_dir), step_dir))
+					SEND_SIGNAL(mob_to_process, COMSIG_OBSTRUCTED_MOVE, step_dir)
+				continue
 			if(prob(stutter_step_prob[mob_to_process]))
-				if(!step(mob_to_process, pick(LeftAndRightOfDir(get_dir(mob_to_process, atoms_to_walk_to[mob_to_process]), diagonal_check = TRUE)))) //Couldn't move, something in the way
-					SEND_SIGNAL(mob_to_process, COMSIG_OBSTRUCTED_MOVE)
+				step_dir = pick(LeftAndRightOfDir(get_dir(mob_to_process, atoms_to_walk_to[mob_to_process])))
+				if(!mob_to_process.Move(get_step(mob_to_process, step_dir), step_dir))
+					SEND_SIGNAL(mob_to_process, COMSIG_OBSTRUCTED_MOVE, step_dir)
 			continue
-
 		if(get_dist(mob_to_process, atoms_to_walk_to[mob_to_process]) < distances_to_maintain[mob_to_process]) //We're too close, back it up
-			if(!step_away(mob_to_process, atoms_to_walk_to[mob_to_process]))
-				SEND_SIGNAL(mob_to_process, COMSIG_OBSTRUCTED_MOVE)
-			mob_to_process.last_move_time = world.time
-			continue
-		if(!step_to(mob_to_process, atoms_to_walk_to[mob_to_process]))
-			SEND_SIGNAL(mob_to_process, COMSIG_OBSTRUCTED_MOVE)
-		mob_to_process.last_move_time = world.time
+			step_dir = get_dir(atoms_to_walk_to[mob_to_process], mob_to_process)
+		else
+			step_dir = get_dir(mob_to_process, atoms_to_walk_to[mob_to_process])
+		var/turf/next_turf = get_step(mob_to_process, step_dir)
+		if(!can_cross_lava_turf(next_turf) || (!mob_to_process.Move(next_turf, step_dir) && !(SEND_SIGNAL(mob_to_process, COMSIG_OBSTRUCTED_MOVE, step_dir) & COMSIG_OBSTACLE_DEALT_WITH)))
+			step_dir = pick(LeftAndRightOfDir(step_dir))
+			next_turf = get_step(mob_to_process, step_dir)
+			if(!can_cross_lava_turf(next_turf))
```

Comment:
this prevent xeno from suiciding on magmoor and barrenq. Might be too costly? i don't really think so